admin/cmd/bootstrap.go

```diff
@@ -211,3 +222,81 @@ func processFinalError(err error, isJSON bool) {
 		os.Exit(1)
 	}
 }
+
+const lookupTimeout = 1 * time.Second
+
+// findSocketOrPort detects if pmm-agent is listening on socket or port.
+// The arguments are used for testing purposes and can be left empty.
+func findSocketOrPort(socketPath string, localPort uint32) (string, uint32) {
+	logrus.Debug("Detecting socket or port of local pmm-agent")
+
+	if socketPath == "" {
+		socketPath = flags.SocketPath
+	}
+
+	if localPort == 0 {
+		localPort = agentlocal.DefaultPMMAgentListenPort
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), lookupTimeout)
+	defer cancel()
+
+	// We make the channels buffered with size 1 to avoid goroutine leaks
+	socketChan := make(chan string, 1)
+	portChan := make(chan uint32, 1)
+
+	// Check socket connection
+	go func() {
+		defer close(socketChan)
+
+		var socket string
+
+		err := checkConnection(ctx, "unix", socketPath)
+		if err == nil {
+			socket = socketPath
+		}
+
+		socketChan <- socket
+	}()
+
+	// Check TCP connection
+	go func() {
+		defer close(portChan)
+
+		var port uint32
+
+		err := checkConnection(ctx, "tcp", net.JoinHostPort(agentlocal.Localhost, strconv.Itoa(int(localPort))))
+		if err == nil {
+			port = localPort
+		}
+
+		portChan <- port
+	}()
+
+	if sock := <-socketChan; sock != "" {
```

Comment:
I assumed we want to use either socket or a port. It looks like this will block until it gets a message on `socketChan`? This means the port check doesn't happen until we have detected a socket, correct?